Make log reading more robust to errors

cpuguy83 opened the pull request:
- fixes / addresses https://github.com/docker/for-linux/issues/140

Fundamentally this change (beyond the minor refactor and tracing spans) is to not treat errors reading from log files as fatal to providing logs to the client.

The need for this change is because often log files may be corrupted either due to old bugs in dockerd or for other reasons.
One specific thing we have seen seems to be caused by ext4 by default is not doing full journaling. On power loss or kernel panic this can cause the log data to be corrupted.
We see this mostly on IoT devices which are often not on battery backup or frequently hard cut from power.

Instead of erroring out a stream on errors from a particular file this just skips on to the next file.

I had made a previous attempt to be a little more clever and scrub the file to try and find a place where we can start parsing again but realized this can cause serious issues and is generally error prone.

This also takes care of a long-standing todo to handle log files lazily (or at least as lazily as possible) so we only end up decompressing files just-in-time.

Unfortunately we don't have a way to return warnings to the client that we had to skip files, however with the tracing changes at least it could be seen there and in the daemon logs.

```markdown changelog
When reading logs with the `jsonfile` or `local` log drivers, any errors while trying to read or parse underlying log files will cause the rest of the file to be skipped and move to the next log file (if one exists) rather than returning an error to the client and closing the stream.
The errors are viewable in the dockerd logs and exported to traces when tracing is configured.

When reading log files, compressed log files are now only decompressed when needed rather than decompressing all files before starting the log stream.
```

cpuguy83 committed:
Refactor logfile reads

This simplifies how we manage log files, especially rotated ones.
It also fixes a long-standing issue to lazily open rotated files so we
don't needlessly start decompressing files that we don't need.

Much of this is just setting things up for commits following this one.

It uses ReaderAtSize for managing all files to be tailed and manages
cleanups by passing closures.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Logfile: skip files that are corrupted

When there is an error in parsing an individual log file just close the
log and move on to the next one instead of erroring our the entire
request.

I investigated trying to error correct and scan ahead for corrupted log
files but found this is too much of a risk of parsing things we
shouldn't be and hence why this is just dropping the rest of the file.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 committed:
Logfile: Add tracing spans

This plumbs a context down the stack and handles cancellation as needed
so that we can have correlated traces from the API.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
```
daemon/logger/loggerutils/logfile.go:938:2: unreachable: unreachable code (govet)
	return nil
```

Unreachable code, or code that simply does not exist?
There's only 654 LOC in this file and linter is not failing locally.

cpuguy83 commented:
The issue with Windows seems to be caused by the frequency with which time is sampled on Windows, which is by default every 15ms.
We can request greater precision from windows (e.g. 1ms which is what linux has) but this has performance implications.
Something I could do is get the time the file was originally rotated off the main log file and request the higher frequency time at that point and store that in the gzip header. This would mean we are requesting the higher frequency only on rotate and not on every log message.

It does bring up the point that relying on time for determining if a decompressed file relates to a particular compressed file is not completely reliable, but 1ms frequency seems like its "good enough" for this case.

Another alternative would be to digest the original log file ands tore that in the gzip header.

cpuguy83 commented on pkg/tailfile/tailfile.go:
```diff
@@ -71,7 +71,7 @@ func NewTailReaderWithDelimiter(ctx context.Context, r SizeReaderAt, reqLines in
 
 	if int64(len(delimiter)) >= size {
-		return bytes.NewReader(nil), 0, nil
+		return io.NewSectionReader(bytes.NewReader(nil), 0, 0), 0, nil
```
Need to revert these back

cpuguy83 commented on daemon/logger/loggerutils/sharedtemp.go:
```diff
@@ -192,6 +193,8 @@ type sharedFileReader struct {
 	closed bool
 }
 
+var _ io.Closer = (*sharedFileReader)(nil)
```
Remove this

cpuguy83 commented:
Ok I think this is ready.
@corhere I put back sharedtemp as discussed.

corhere reviewed (commented):
I have noticed that the refactor also makes it possible for log drivers to implement "skip to next file on decode failure" behavior themselves, simply by having their `(loggerutils.Decoder).Decode()` implementation return `io.EOF` instead of the JSON decode error or whatever. And drivers which use a self-synchronizing stream format could implement more intelligent recovery logic since the decoder is in charge of advancing the reader. But I have convinced myself that it is still worthwhile to lift the "skip file" logic into `loggerutils` because it works as a fallback for all log drivers, including hypothetical drivers with intelligent recovery logic. It'll be there as a fallback recovery mechanism if the driver's resynchronization logic falls over (returns an unexpected error) on a particularly corrupt file.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -587,27 +785,48 @@ func tailFiles(files []SizeReaderAt, watcher *logger.LogWatcher, dec Decoder, ge
+		}
+		idx++
+
+		if err != nil {
+			if errors.Is(err, errConsumerGone) {
+				span.AddEvent("consumer gone")
+				return false
+			}
+
+			if errors.Is(err, errLogTooNew) {
+				span.AddEvent("Remaining logs are newer than requested time range")
+				return false
+			}
+			log.G(ctx).WithError(err).Warn("Skipping remainder of log file due to error")
 		}
```
This works for me.

I _think_ the reason I originally did this was because the decoder was returning an error with the bad offset, then we'd advance that offset in logfile.
This is of course not part of it anymore and as you pointed out we can probably deal with those type of errors directly in the decoder anyway.

cpuguy83 committed:
Fix time comparison in sharedtemp implementation

Times cannot be compared with `==` and instead should use the `t.Equal`
function.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

corhere commented:
```
=== FAIL: github.com/docker/docker/daemon/logger/loggerutils TestTailFiles/handle_corrupted_data (10.01s)
    logfile_test.go:189: timeout waiting for log message
```

cpuguy83 commented:
So unfortunately its not _that_ straight forward to have forwarder manage everything and still return just a bool because logfile following needs to have different error handling semantics from tailing.

What I ended p doing here is wrapping the decoder's `Decode` function from `tailFiles` to make it so if the underlying decoder returns an error then the error is wrapped in an `io.EOF`, which allows for this different error handling semantics.

corhere committed:
logger/journald: plumb contexts into reader

Co-authored-by:: Cory Snider <csnider@mirantis.com>
Signed-off-by: Brian Goff <cpuguy83@gmail.com>

corhere commented on daemon/logger/journald/read.go:
```diff
@@ -448,7 +450,7 @@ func (r *reader) signalReady() {
 }
 
-func (s *journald) ReadLogs(config logger.ReadConfig) *logger.LogWatcher {
+func (s *journald) ReadLogs(ctx context.Context, config logger.ReadConfig) *logger.LogWatcher {
```
Plumbing contexts fully into the journald reader with cancelation is pretty straightforward: https://github.com/corhere/moby/commit/b603933dc10b8873455fe8c2d7489b23aa67d892
Feel free to cherry-pick that commit into this PR. Otherwise I would be happy to rebase and open a follow-up PR once this one has landed.

cpuguy83 replied:
Pulled in that commit (with some minor changes).

corhere commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -389,19 +393,104 @@ func (w *LogFile) ReadLogs(config logger.ReadConfig) *logger.LogWatcher {
+	w.fsopMu.RUnlock()
+
+	if err != nil {
+		// TODO: SHould we allow this to continue (as in set `cont=true`) and not error out the log stream?
```
```suggestion
		// TODO: Should we allow this to continue (as in set `cont=true`) and not error out the log stream?
```

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -587,27 +784,58 @@ func tailFiles(files []SizeReaderAt, watcher *logger.LogWatcher, dec Decoder, ge
+		return func() (*logger.Message, error) {
+			msg, err := dec.Decode()
+			if err != nil && !errors.Is(err, io.EOF) {
+				err = fmt.Errorf("%w: %w", io.EOF, err)
```
Decoders don't have control over underlying I/O though so I'm not really sure they can be cancellable.
What we can do is start a goroutine waiting for the parent to cancel and then close the file, which should trigger the decode to return EOF/UnexpectedEOF.

corhere replied:
A robust decoder for a self-synchronizing log format could test if the context is done between the `Read()`s to scan for the next record delimiter and exit the loop early. That said, I like your idea better as it would work unmodified with any decoder implementation that handles `fs.ErrClosed` by returning.

cpuguy83 committed:
logfile: Close reader when caller cancels

This allows for an individual decode operation to be cancelled while the
log reader is reading data from a log file by closing the underlying file.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>

cpuguy83 commented:
@corhere I think I've addressed everything here.

corhere commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -813,6 +814,17 @@ func tailFiles(ctx context.Context, files []fileOpener, watcher *logger.LogWatch
+		// can close the underlying reader while a the decoder may be waiting for
+		// data. This should trigger an early return from the decoder.
+		ctx, cancel := context.WithCancel(ctx)
+		go func() {
+			<-ctx.Done()
+			if err := ra.Close(); err != nil {
+				log.G(ctx).WithError(err).Debug("Error closing log reader")
+			}
+		}()
```
Since our minimum supported Go version is go1.21, we can use [`context.AfterFunc`](https://pkg.go.dev/context#AfterFunc) so we don't need to keep a goroutine parked for the entirety of the tailing operation.
```suggestion
		context.AfterFunc(ctx, func() {
			if err := ra.Close(); err != nil {
				log.G(ctx).WithError(err).Debug("Error closing log reader")
			}
		})

cpuguy83 replied:
I certainly can, but `context.AfterFunc` seems incredibly overkill and still has a parked goroutine.

cpuguy83 replied:
In fact, I think `context.AfterFunc` is entirely the wrong solution here, as it seems the entire point is being able to know when the returned cancel function actually cancels.

corhere replied:
Huh? Knowing when the context is done without parking a goroutine is the entire rationale for `context.AfterFunc`! https://go.dev/issue/57928

cpuguy83 replied:
`contxt.AfterFunc` is literally parking a goroutine and doing a bunch of other internal state management.

corhere replied:
https://github.com/golang/go/commit/54d429999c00f277c80b5698fb78c7b71f91aad9

> Add support for contexts that implement an AfterFunc method, which
> can be used to avoid the need to start a new goroutine watching
> the Done channel when propagating cancellation signals.

cpuguy83 replied:
So your suggestion here is to implement an afterFunctioner context?

cpuguy83 replied:
For posterity, we discussed this on a maintainers call and the underlying context implementation has an internal optimization that it can use here such that it won't spin up a goroutine.

cpuguy83 replied:
I've updated this to use context.AfterFunc

vvoland commented on daemon/logger/loggerutils/logfile_test.go:
```diff
@@ -16,91 +18,179 @@ import (
+		files := makeOpener(bytes.NewReader(f1.Bytes()), bytes.NewReader(f2.Bytes()), bytes.NewReader(f3.Bytes()))
+
+		// At this point we our log "files" should have 4 log messages in it
+		// intersperesed with some junk that is invalid json
```
```suggestion
		// interspersed with some junk that is invalid json
```

cpuguy83 replied:
Fixed.

vvoland reviewed (commented):
Some surficial comments

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -468,117 +517,265 @@ func (w *LogFile) readLogsLocked(currentPos logPos, config logger.ReadConfig, wa
+		span.SetAttributes(attribute.Stringer("lastLogTime", extra.LastTime))
+	}
+
+	span.AddEvent("Start decompress")
```
I considered this but find that the entries in the daemon logs with this are largely not useful.
I could log this as `Trace` level but then we get nothing in the span since trace logging isn't enable in the logrus logger.

Note, I created that tracing hook 😄

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -622,16 +861,35 @@ func newForwarder(config logger.ReadConfig) *forwarder {
+			span.AddEvent("watch consumer gone")
+			return false
+		case <-ctx.Done():
+			span.AddEvent(ctx.Err().Error())
```
I think the reason I did this was because its not really an error that the caller cancelled the context... but am fine either way.

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -468,117 +517,265 @@ func (w *LogFile) readLogsLocked(currentPos logPos, config logger.ReadConfig, wa
-				f.Close()
+		if retErr != nil {
+			for _, fo := range out {
+				fo.Close()
```
There is no error return here, but indeed it is ignoring errors within the `Close` calls.

vvoland replied:
Oh, right 👍🏻

cpuguy83 commented on daemon/logger/loggerutils/logfile.go:
```diff
@@ -468,117 +517,265 @@ func (w *LogFile) readLogsLocked(currentPos logPos, config logger.ReadConfig, wa
+	}
+	defer gzr.Close()
+
+	// Extract the last log entry timestramp from the gzip header
```
Fixed.

vvoland commented:
Hmmm, interesting panic:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x556232a85fdd]

goroutine 57732 [running]:
github.com/docker/docker/daemon/containerd.(*ImageService).imageSummary.func1.2(0xc0003c9b80?)
	/go/src/github.com/docker/docker/daemon/containerd/image_list.go:358 +0xfd
github.com/docker/docker/container.(*memoryStore).ApplyAll.func1(0x0?)
	/go/src/github.com/docker/docker/container/memory_store.go:77 +0x23
created by github.com/docker/docker/container.(*memoryStore).ApplyAll in goroutine 57632
	/go/src/github.com/docker/docker/container/memory_store.go:76 +0x52
[rootlesskit:child ] error: command [/usr/local/bin/dockerd-rootless.sh --debug --host unix:///go/src/github.com/docker/docker/bundles/test-integration/docker.sock --storage-driver overlayfs --pidfile bundles/test-integration/docker.pid --userland-proxy=true --experimental] exited: exit status 2
```


Unrelated to this PR, but what's strange here this stack trace matches the code from `master`, but this PR is based off a `master` where this code is not present yet.

> github.com/docker/docker/daemon/containerd.(*ImageService).imageSummary.func1.2(0xc0003c9b80?)
>	/go/src/github.com/docker/docker/daemon/containerd/image_list.go:358 +0xfd

Code from this PR:
https://github.com/moby/moby/blob/6d941222173a11ed25c3a5daa75c21e03fe28ea4/daemon/containerd/image_list.go#L356-L358

Code from master:
https://github.com/moby/moby/blob/d8f079deb6d9faf6540b756baec42ede172cd53e/daemon/containerd/image_list.go#L356-L358

vvoland reviewed (approved):
LGTM; but please try to add a changelog description 😅

cpuguy83 commented:
Added changelog description.

cpuguy83 commented:
Opened backport PR here: https://github.com/moby/moby/pull/48842

thaJeztah commented:
Cross-referencing, in case this patch would make its way in other branches;

- https://github.com/moby/moby/issues/49075
- https://github.com/moby/moby/pull/49078
